File: Hashing/hash_table_rehashing.py
# File Name: hash_table_rehashing.py

class HashTable:
    def __init__(self, size=5):
        self.size = size
        self.count = 0
        self.table = [None] * size
        self.LOAD_FACTOR = 0.7

    def hash_function(self, key):
        return key % self.size

    def rehash(self):
        old_table = self.table
        self.size *= 2
        self.table = [None] * self.size
        self.count = 0

        for key in old_table:
            if key is not None:
                self.insert(key)

        print(f"Rehashing done. New size = {self.size}")

    def insert(self, key):
        if self.count / self.size > self.LOAD_FACTOR:
            self.rehash()

        index = self.hash_function(key)

        while self.table[index] is not None:
            index = (index + 1) % self.size

        self.table[index] = key
        self.count += 1
        print(f"Key {key} inserted")

    def display(self):
        print("\nHash Table:")
        for i in range(self.size):
            if self.table[i] is None:
                print(f"Index {i} -> EMPTY")
            else:
                print(f"Index {i} -> {self.table[i]}")


if __name__ == "__main__":
    ht = HashTable()

    while True:
        print("\n1.Insert\n2.Display\n3.Exit")
        choice = int(input("Enter choice: "))

        if choice == 1:
            ht.insert(int(input("Enter key: ")))
        elif choice == 2:
            ht.display()
        elif choice == 3:
            break

File: Hashing/hash_table_rehashing.c
/* File Name: hash_table_rehashing.c */

#include <stdio.h>
#include <stdlib.h>

#define INITIAL_SIZE 5
#define LOAD_FACTOR 0.7
#define EMPTY -1

int *hashTable;
int size = INITIAL_SIZE;
int count = 0;

// Hash Function
int hashFunction(int key) {
    return key % size;
}

// Rehash Function
void rehash() {
    int oldSize = size;
    int *oldTable = hashTable;

    size = size * 2;
    hashTable = (int *)malloc(size * sizeof(int));

    for (int i = 0; i < size; i++)
        hashTable[i] = EMPTY;

    count = 0;

    for (int i = 0; i < oldSize; i++) {
        if (oldTable[i] != EMPTY) {
            int key = oldTable[i];
            int index = hashFunction(key);

            while (hashTable[index] != EMPTY)
                index = (index + 1) % size;

            hashTable[index] = key;
            count++;
        }
    }

    free(oldTable);
    printf("Rehashing done. New size = %d\n", size);
}

// Insert
void insert(int key) {
    float load = (float)count / size;

    if (load > LOAD_FACTOR)
        rehash();

    int index = hashFunction(key);

    while (hashTable[index] != EMPTY)
        index = (index + 1) % size;

    hashTable[index] = key;
    count++;
    printf("Key %d inserted\n", key);
}

// Display
void display() {
    printf("\nHash Table:\n");
    for (int i = 0; i < size; i++) {
        if (hashTable[i] == EMPTY)
            printf("Index %d -> EMPTY\n", i);
        else
            printf("Index %d -> %d\n", i, hashTable[i]);
    }
}

// Main
int main() {
    hashTable = (int *)malloc(size * sizeof(int));
    for (int i = 0; i < size; i++)
        hashTable[i] = EMPTY;

    int choice, key;

    while (1) {
        printf("\n1.Insert\n2.Display\n3.Exit\n");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter key: ");
                scanf("%d", &key);
                insert(key);
                break;
            case 2:
                display();
                break;
            case 3:
                exit(0);
        }
    }
}
